KAFKA-16016: Add docker wrapper in core and remove docker utility script

VedarthConfluent opened the pull request:
Migrates functionality provided by utility to Kafka core. This wrapper will be used to generate property files and format storage when invoked from docker container.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

kagarwal06 committed:
KAFKA-16016: Add boilerplate code for the docker wrapper

VedarthConfluent committed:
KAFKA-16016: Migrate ensure and path to bash

VedarthConfluent committed:
Add server properties generation logic

VedarthConfluent committed:
Add log4j and tools support

VedarthConfluent committed:
Fix bug

VedarthConfluent committed:
Fix bash script related issues

VedarthConfluent committed:
Run all shell scripts in a single shell

kagarwal06 committed:
KAFKA-16016: Refactor and add scalastyle code

VedarthConfluent committed:
Refactor server properties preparation

VedarthConfluent committed:
Fix build error

kagarwal06 committed:
KAFKA-16016: Add scalastyle code

VedarthConfluent committed:
Migrate all file operations to scala

VedarthConfluent committed:
Refactor changes

kagarwal06 committed:
KAFKA-16016: Refactor code and rename variables

kagarwal06 committed:
KAFKA-16016: Refactor to make code more testable

kagarwal06 committed:
KAFKA-16016: Add first UT

VedarthConfluent committed:
Add test for system properties environment variable parsing

VedarthConfluent committed:
Add test for prepareServerConfig and prepareToolsLog4jConfigs

kagarwal06 committed:
KAFKA-16016: Add more UTs

VedarthConfluent committed:
Add log4j behavioural tests

kagarwal06 committed:
KAFKA-16016: Error handling

VedarthConfluent committed:
Code cleanup and fix launch script

VedarthConfluent committed:
Delete redundant files

VedarthConfluent committed:
Remove utility from Dockerfile

VedarthConfluent committed:
Merge branch 'trunk' of github.com:apache/kafka into krish-docker-wrapper-KAFKA-16016

VedarthConfluent committed:
Remove golang utility code

mimaison commented on docker/jvm/launch:
```diff
@@ -28,25 +27,25 @@ fi
+export KAFKA_JVM_PERFORMANCE_OPTS="${KAFKA_JVM_PERFORMANCE_OPTS-} -XX:SharedArchiveFile=/opt/kafka/storage.jsa"
 
 echo "===> Using provided cluster id $CLUSTER_ID ..."
 # A bit of a hack to not error out if the storage is already formatted. Need storage-tool to support this
```
Could we handle that in the wrapper?

VedarthConfluent replied:
Sounds good. Used `--ignored-formatted` flag of storage format command inside the wrapper. This will allow us to restart the docker container without running into an error, and thus remove the need for this hack

VedarthConfluent replied:
Update on this, I have moved this check back to the bash script, given that the documentation suggests that `--ignored-formatted` flag should only be used in a very specific scenario only and I couldn't find another way to handle it in the wrapper. Maybe it will be feasible once storage tool is rewritten.
Given that this refactor was mentioned as a minor concern, hopefully it should be okay if we keep this small logic in bash script for now.

mimaison reviewed (commented):
Thanks for the PR, It looks good!
I left a small question. I think it shouldn't be too hard to do but if for some reasons there are complications we can probably merge as is.

VedarthConfluent committed:
Use ignore formatted flag to avoid errors on container restart

VedarthConfluent committed:
Update the inline documentation for the launch script

VedarthConfluent committed:
Fix formatting

viktorsomogyi commented on core/src/main/scala/kafka/docker/KafkaDockerWrapper.scala:
```diff
@@ -0,0 +1,218 @@
+      throw new RuntimeException(s"Error: No operation input provided. " +
+        s"Please provide a valid operation: 'setup'.")
+    }
+    val operation = args.head
```
While this is a simple command and it works, it would be more future proof to use argparse4j as do tools.

VedarthConfluent replied:
Fair point. Made the required changes

viktorsomogyi commented on core/src/main/scala/kafka/docker/KafkaDockerWrapper.scala:
```diff
@@ -0,0 +1,218 @@
+        }
+        val defaultConfigsDir = arguments(0)
+        val mountedConfigsDir = arguments(1)
+        val finalConfigsDir = arguments(2)
```
Parsing these 3 arguments would also be a good opportunity to validate and convert these into `Path` objects. Using the API of `Path` also makes your code more robust down the line when you append the config file names in `prepareConfigs`.

VedarthConfluent replied:
Thanks for the suggestion. Incorporated

viktorsomogyi commented on core/src/main/scala/kafka/docker/KafkaDockerWrapper.scala:
```diff
@@ -0,0 +1,218 @@
+
+  import Constants._
+
+  private def formatStorageCmd(configsDir: String, env: Map[String, String]): Array[String] = {
```
While I agree that it works, it's not a nice solution. Would it be a big refactor to use a specific method from `StorageTool` instead of `main`?

VedarthConfluent replied:
The way storage tool is written, we will need the logic that's present in the main function. We can create a new function that contains entirety of the main function logic and call that from both storage tool main and our code. But that will be kind of same as the present solution, but with just one extra method. 
Proper solution will require thorough refactoring of the storage tool, which I think is out of scope for this change. Please let us know your thoughts on this.

mimaison replied:
`StorageTool` is being rewritten in Java in https://github.com/apache/kafka/pull/14847 so we can probably keep the code as is for now and see if we can tidy things up once the rewrite PR is merged.

viktorsomogyi replied:
I'm OK with that (refactoring later).

omkreddy replied:
Can we create JIRA to track this?

VedarthConfluent replied:
Done! https://issues.apache.org/jira/browse/KAFKA-16090

VedarthConfluent committed:
Refactor changes

- Use argparse4j for parsing arguments
- Use path across the wrapper to make the code more robust

mimaison commented:
Is there a way to test this in the Apache CI or Github action? 

On my laptop `docker_build_test.py` fails due to some networking errors like ` Connection to node -1 (localhost/127.0.0.1:9092) could not be established.`. I've not had time to investigate it but it would be nice to perform checks in the CI.

VedarthConfluent commented:
@mimaison There is a GitHub Actions Workflow added for just that! You can check it out [here](https://github.com/apache/kafka/actions/workflows/docker_build_and_test.yml). Also there is helpful documentation for building, testing and releasing the docker image [here](https://github.com/apache/kafka/blob/trunk/docker/README.md)

kagarwal06 commented:
> I saw that but the action only allows me to pick a branch or tag. Ideally I'd like to run it on the pull request.

Hi @mimaison , thanks for the feedback.

The dockerisation of Apache Kafka follows the following process
- Accepts a signed Kafka tarball link.
- Expects the presence of both the `.asc` file and `key` for tarball verification. This verification process is implemented in the [Dockerfile](https://github.com/apache/kafka/blob/trunk/docker/jvm/Dockerfile#L78-L83).

The `docker_build_test.py`, utilized in GitHub Actions as well, takes an Apache Kafka tarball URL, supplies it to the Dockerfile, builds the Docker image, and runs tests on it. This approach is valuable during Release Candidate (RC) testing, allowing us to verify the Docker image by providing the Apache Kafka RC link along with the necessary `.asc` file and the `key`.

However, in the context of pull requests, especially when changes impact the Apache Kafka codebase and the docker path (as in the current PR), testing requires a signed tarball with the associated `keys` and `.asc` file. Presently, our pipeline is not designed for this use-case.

**NOTE:** We have locally tested the above changes by building the Apache Kafka and uploading the tarball to S3 and commenting out the tarball verification code in the Dockerfile.

VedarthConfluent committed:
Move the already formatted check back to bash script

omkreddy commented:
> I saw that but the action only allows me to pick a branch or tag. Ideally I'd like to run it on the pull request.

@VedarthConfluent Can you pls crate JIRA to track this?

omkreddy reviewed (approved):
@VedarthConfluent Thanks for the LGTM. I have done few manual tests.
 
As @mimaison mentioned, we need to have option of build and test docker image on PR request.

I think its good have to this change in 3.7 (so that we can avoid Go script in 3.7). If there are no concerns, i would like to cherry-pick this change to 3.7 branch.

cc @stanislavkozlovski

VedarthConfluent commented:
I have created a jira to track the new github actions pipeline requested above https://issues.apache.org/jira/browse/KAFKA-16091

mimaison commented:
@omkreddy Does this work for you?

1. Checkout the branch from the PR and go to the `docker` directory
2. Run `python3 docker_build_test.py kafka/test --image-tag=3.6.0 --image-type=jvm --kafka-url=https://archive.apache.org/dist/kafka/3.6.0/kafka_2.13-3.6.0.tgz`

omkreddy commented:
> @omkreddy Does this work for you?
> 
> 1. Checkout the branch from the PR and go to the `docker` directory
> 2. Run `python3 docker_build_test.py kafka/test --image-tag=3.6.0 --image-type=jvm --kafka-url=https://archive.apache.org/dist/kafka/3.6.0/kafka_2.13-3.6.0.tgz`

Hi @mimaison, It will not work now, as the newly added `KafkaDockerWrapper.scala` classes are not part of 3.6. 
For now,  I have tested with custom image and few changes to the PR. We can use above cmd once we have 3.7 RC0. 
As suggested by you, we need docker build option for PRs. I have requested the same  from @VedarthConfluent.

stanislavkozlovski commented:
Let's merge to 3.7
